src/fs_fd.ts

```diff
@@ -162,7 +189,11 @@ export class OpenSyncOPFSFile extends Fd {
     iovs: Array<wasi.Iovec>,
   ): { ret: number; nwritten: number } {
     let nwritten = 0;
-    if (this.file.readonly) return { ret: wasi.ERRNO_BADF, nwritten };
+    if (
+      this.file.readonly ||
+      (Number(this.fs_rights_base) & wasi.RIGHTS_FD_WRITE) === 0
+    )
+      return { ret: wasi.ERRNO_BADF, nwritten };
```

Comment:
For please use `{}` for `if` where the action is not on the same line. That helps prevent issues like the duplicate "goto fail;" that caused a huge security issue in macOS a couple of years back.